### PR TITLE
CMake: Add documentation about CMakeLists.txt file

### PR DESCRIPTION
**Pull request by hugueskamba**
Add documentation to explain how CMakeLists.txt file
are used in Mbed OS.

### CODE REVIEW

File: docs/tools/mbed_cli_2/build-system/cmakelists.md [hunk 1]
```diff
@@ -1 +1,21 @@
-# CMakelists
+# CMakeLists
+
+## Introduction
+CMake allows the description of the project by listing instructions and directives in input source files named `CMakeLists.txt`. Mbed OS has many of these files which are included in the build from a top-level `CMakeLists.txt` at the root of the project.
+We will briefly cover how `CMakeLists.txt` are used within the Mbed OS repository.
+
+## CMake target selection
```
Comment (0xc0170) [inline docs/tools/mbed_cli_2/build-system/cmakelists.md:7]: Should we move info from our docs here?

```
## Naming convention

- `mbed` namespace: Mbed CMake targets must have `mbed-` prefix
- all characters are lower case
- words separated by hyphens

For example: `mbed-ble-cordio`

## Mbed libraries

This section provides description for all CMake targets within Mbed OS repository.

### mbed-os and mbed-baremetal libraries

An application links to either mbed-os or mbed-baremetal.

| Target  | Function |
| ------------- | ------------- |
| mbed-os | Interface library that attaches Mbed OS to a target (objects from mbed-core and mbed-rtos) |
| mbed-baremetal | Interface library that attaches Mbed baremetal to a target (objects from mbed-core) |

These are internal targets. They are subject to change. An application should not link to any of these.

| Target  | Function |
| ------------- | ------------- |
| mbed-cmsis-cortex-m | CMSIS library for Cortex-M |
| mbed-cmsis-cortex-a | CMSIS library for Cortex-A |
| mbed-core  | Contains flags from Mbed baremetal.  All Mbed OS libraries link to this.  |
| mbed-core-objects  | Builds objects for Mbed baremetal  |
| mbed-rtos | Contains flags that turn Mbed baremetal into Mbed OS |
| mbed-rtos-objects | Builds objects for RTOS |
| mbed-os-objects | Build objects for Mbed OS (objects from mbed-core and mbed-rtos) |
| ${MBED_TARGET_CONVERTED} | Contains interface flags and sources that adapt Mbed OS for a target |

### Additional Mbed OS libraries

An application must explicitely link them using `target_link_libraries()`.

| Target  | Function |
| ------------- | ------------- |
| mbed-events  | Events library
| mbed-ble  | Bluetooth library
| mbed-storage  | Storage library
| mbed-mbedtls  | TLS  library
| mbed-nfc  | NFC library
| mbed-nanostack  | Nanostack library
| mbed-netsocket | Netsocket library
| mbed-lorawan | Lorawan library
| mbed-wifi | Wifi library

```

File: docs/tools/mbed_cli_2/build-system/cmakelists.md [hunk 2]
```diff
@@ -1 +1,21 @@
...
+
+Mbed OS is built as a collection of all interdependent libraries essential to run the most basic application. This collection of library is used to create the CMake library target aptly named `mbed-core`. `mbed-core` is the foundation library which is linked with other libraries to make more complex systems. Applications that require an RTOS should link with the `mbed-os` CMake library target which includes `mbed-core` as well as the library that implements the RTOS. Applications destined to run on hardware with greater memory constrains and do not require an RTOS should use the `mbed-baremetal` CMake library instead of `mbed-os`.
```
Comment (0xc0170) [inline docs/tools/mbed_cli_2/build-system/cmakelists.md:9]: `mbed-core` should be used to get target/MCU/cmsis and other details to custom libraries (get flags basically) - should this be in here ?

File: docs/tools/mbed_cli_2/build-system/cmakelists.md [hunk 3]
```diff
@@ -1 +1,21 @@
...
+All Mbed CMake libraries are prefixed with `mbed-`.
+
+The CMake targets defined in the `CMakeLists.txt` of a subdirectory must be explicitly built as they are excluded from the `ALL` of the parent directory. This is because we add subdirectories using `EXCLUDE_FROM_ALL` so only the features and BSP relevant to a project and Mbed boards are excluded in the build unless needed.
+
+## Library configuration
+Source files are included in libraries by listing them in `CMakeLists.txt` files using `target_sources()`. However, header files are included by listing the directories where they can be found using `target_include_directories()`. Additionally, compile defintions can also de added to libraries using `target_compile_defintions()`
+
+## Board support
+All Mbed boards can be built with CMake. An Mbed board specific configuration CMake module is generated by Mbed CLI 2 to help in the selection of the board support package. See Mbed CLI 2 for more information about the generated CMake module.
+Post build operations required for some Mbed boards are triggered in CMake by invoking the appropriate script.
```
Comment (0xc0170) [inline docs/tools/mbed_cli_2/build-system/cmakelists.md:22]: post build op will be in detail described in porting so here only , should we add a reference to it, or fine as it is?

File: docs/tools/mbed_cli_2/build-system/cmakelists.md [hunk 4]
```diff
@@ -1 +1,21 @@
...
+
+Other libraries are optional and can be linked with the application CMake target as required. To identify the name of the optional library to link with, inspect the CMake input source file (`CMakeLists.txt`) in the directory containing the source files of the library required. For example if LoRaWAN is required, inspect `connectivity/lorawan/CMakeLists.txt` and you see that the CMake library which implements it is `mbed-lorawan`.
+All Mbed CMake libraries are prefixed with `mbed-`.
+
+The CMake targets defined in the `CMakeLists.txt` of a subdirectory must be explicitly built as they are excluded from the `ALL` of the parent directory. This is because we add subdirectories using `EXCLUDE_FROM_ALL` so only the features and BSP relevant to a project and Mbed boards are excluded in the build unless needed.
+
+## Library configuration
+Source files are included in libraries by listing them in `CMakeLists.txt` files using `target_sources()`. However, header files are included by listing the directories where they can be found using `target_include_directories()`. Additionally, compile defintions can also de added to libraries using `target_compile_defintions()`
```
Comment (0xc0170) [inline docs/tools/mbed_cli_2/build-system/cmakelists.md]: `target_link_libraries()` to link targets together or add dependency?

File: docs/tools/mbed_cli_2/build-system/cmakelists.md (hunk 1 above)
Comment (hugueskamba) [reply docs/tools/mbed_cli_2/build-system/cmakelists.md:7]: I think the different libraries available should be a separate page with more descriptive definition for each library.

File: docs/tools/mbed_cli_2/build-system/cmakelists.md (hunk 2 above)
Comment (hugueskamba) [reply docs/tools/mbed_cli_2/build-system/cmakelists.md:9]: I thought this page was meant to not have too many details like that. I thought it was just meant to describe how CMakeLists.txt are used.
What you are talking about here is details about a library which IMO should be somewhere else where we talk about the libraries available.

File: docs/tools/mbed_cli_2/build-system/cmakelists.md (hunk 3 above)
Comment (hugueskamba) [reply docs/tools/mbed_cli_2/build-system/cmakelists.md:22]: I do not think it makes sense to reference that page here.

File: docs/tools/mbed_cli_2/build-system/cmakelists.md (hunk 4 above)
Comment (hugueskamba) [reply docs/tools/mbed_cli_2/build-system/cmakelists.md]: Yes, I can add that.

File: docs/tools/mbed_cli_2/build-system/cmakelists.md (hunk 1 above)
Comment (0xc0170) [reply docs/tools/mbed_cli_2/build-system/cmakelists.md:7]: We will need that but probably not in the build system section but rather in introduction or somewhere. I'll create a story for this.

File: docs/tools/mbed_cli_2/build-system/cmakelists.md (hunk 2 above)
Comment (0xc0170) [reply docs/tools/mbed_cli_2/build-system/cmakelists.md:9]: true, we will need libraries.md file

File: docs/tools/mbed_cli_2/build-system/cmakelists.md (hunk 4 above)
Comment (hugueskamba) [reply docs/tools/mbed_cli_2/build-system/cmakelists.md]: [Done](https://github.com/ARMmbed/mbed-os-5-docs/compare/28c9170b3469fe9bbcbc353ebd68d3281401d1b8..1eee1390f710e4cf656ca3462fbc74541cb78050).

File: docs/tools/mbed_cli_2/build-system/cmakelists.md [hunk 5]
```diff
@@ -1 +1,22 @@
...
+All Mbed CMake libraries are prefixed with `mbed-`.
+
+The CMake targets defined in the `CMakeLists.txt` of a subdirectory must be explicitly built as they are excluded from the `ALL` of the parent directory. This is because we add subdirectories using `EXCLUDE_FROM_ALL` so only the features and BSP relevant to a project and Mbed boards are excluded in the build unless needed.
+
+## Library configuration
+Source files are included in libraries by listing them in `CMakeLists.txt` files using `target_sources()`. Some source code are provided in Mbed as pre-built libraries and are added to libraries using `target_link_libraries()`. `target_link_libraries()` is also used to link other Mbed libraries if there are dependencies.
+Header files are included by listing the directories where they are found using `target_include_directories()`. Additionally, compile defintions are added to libraries using `target_compile_defintions()`.
```
Comment (0xc0170) [inline docs/tools/mbed_cli_2/build-system/cmakelists.md]:
```suggestion
compile defintions are added to libraries using `target_compile_definitions()`.
```

**Review by 0xc0170 — APPROVED** [review]
LGTM once fixed

File: docs/tools/mbed_cli_2/build-system/cmakelists.md (hunk 5 above)
Comment (hugueskamba) [reply docs/tools/mbed_cli_2/build-system/cmakelists.md]: I do not understand the suggestion. Are you suggesting to remove talking about `target_include_directories()`? If so, why?

Comment (0xc0170) [reply docs/tools/mbed_cli_2/build-system/cmakelists.md]: `defintions` typo

Comment (0xc0170) [reply docs/tools/mbed_cli_2/build-system/cmakelists.md]: The suggestion was incorrect 🤦

Comment (hugueskamba) [reply docs/tools/mbed_cli_2/build-system/cmakelists.md]: [Done](https://github.com/ARMmbed/mbed-os-5-docs/compare/1eee1390f710e4cf656ca3462fbc74541cb78050..0cd3f1dd5233e9248669e14c2c49e5176ccb589e).

**Review by 0xc0170 — APPROVED** [review]

**Pull request merged**